mlir/lib/Dialect/Rock/Transforms/ShuffleGemmForReductions.cpp

```diff
@@ -0,0 +1,597 @@
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/utility/builderUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKSHUFFLEGEMMFORREDUCTIONS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-shuffle-gemm-for-reductions"
+
+using namespace mlir;
+using namespace mlir::arith;
+using namespace mlir::rock;
+
+namespace {
+struct RockShuffleGemmForReductionsPass
+    : public rock::impl::RockShuffleGemmForReductionsBase<
+          RockShuffleGemmForReductionsPass> {
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+ArrayAttr reverse(ArrayAttr attrs) {
+  SmallVector<Attribute> attrsReversed = llvm::to_vector(llvm::reverse(attrs));
+  IRRewriter rewriter(attrs.getContext());
+  return rewriter.getArrayAttr(attrsReversed);
+}
+
+ArrayAttr getAllViewsFromSource(OpOperand *operand) {
+  Value val = operand->get();
+  SmallVector<Attribute> attrs;
+  while (TransformOp trOp = dyn_cast<TransformOp>(val.getDefiningOp())) {
+    attrs.push_back(trOp.getTransformAttr());
+    val = trOp.getViewSource();
+  }
+  IRRewriter rewriter(val.getContext());
+  ArrayAttr arrAttrs = rewriter.getArrayAttr({attrs});
+  return reverse(arrAttrs);
+}
+
+FailureOr<std::tuple<ArrayAttr, Operation *>>
+obtainViewsFromReaderToWriter(memref::AllocOp buffer,
+                              const BufferDependencyAnalysis &deps,
+                              ArrayAttr currViews) {
+  LLVM_DEBUG(llvm::dbgs() << "buffer = " << buffer << "\n");
+  IRRewriter rewriter(buffer.getContext());
+  std::optional<llvm::SmallVector<OpOperand *>> writersOperands =
+      deps.getWriters(buffer);
+  if (!writersOperands.has_value())
+    return failure();
+  for (OpOperand *writerOperand : writersOperands.value()) {
+    ArrayAttr viewsFromAllocOp = getAllViewsFromSource(writerOperand);
+    currViews = prependUpperViews(rewriter, currViews, viewsFromAllocOp);
+    if (isa<GridwiseGemmAccelOp, GridwiseGemmOp>(writerOperand->getOwner())) {
+      return std::make_tuple(reverse(currViews), writerOperand->getOwner());
+    }
+    LLVM_DEBUG(llvm::dbgs()
+               << "write op = " << *writerOperand->getOwner() << "\n");
+    auto writeOp = dyn_cast<MemoryEffectOpInterface>(writerOperand->getOwner());
+    if (!writeOp) {
+      LLVM_DEBUG(llvm::dbgs() << "\tit is not a memory effect interface op\n");
+      continue;
+    }
+    SmallVector<MemoryEffects::EffectInstance> effects;
+    writeOp.getEffects(effects);
+    for (const MemoryEffects::EffectInstance &effect : effects) {
+      OpOperand *readOperand = effect.getEffectValue<OpOperand *>();
+      LLVM_DEBUG(llvm::dbgs()
+                 << "readOperand = " << readOperand->get() << "\n");
+      // Test against the write operand to guard against [MemRead, MemWrite]
+      if (readOperand && readOperand != writerOperand &&
```

Comment:
Doesn't `BufferDependencyAnalysis` do this?

Reply:
Or have this cached?

Reply:
This is heavily inspired by your code here : https://github.com/ROCm/rocMLIR/blob/82378ac44b8627ecbaf9078353fb53588090fe28/mlir/lib/Dialect/Rock/Transforms/Regularize.cpp#L282-L290.

I think its safer to have here as well.